### DISCUSSION TITLE
Making the right kind of installer / setup

### DISCUSSION THREAD

**Discussion by tpitman** [cefsharp/CefSharp#3682] [General]
I have created a very simple browser wrapper using CefSharp to access a specific website I have created.

I now need to make a general installer to deploy this app. I tried a ClickOnce and it installs without errors including the .NET core 3.1 package. It doesn't do anything when I tried to run it. It just shows the hour glass for a second and then nothing. No errors or anything.

The weird thing is that if I find the .exe that is installed and click on it directly everything works.

So I then tried creating a Visual Studio Windows Setup MSI project.

I changed the Publish output so it just used the Folder method with NO ClickOnce.

If I go to the publish output under netcoreapp3.1\publish and double click the exe there on my dev computer, it runs fine.

If I then build the MSI setup project that contains the publish output as the input for it I get the MSI.

When I install that I end up with a C:\Program Files\[companyname]\[appname] folder that has all of the same files contained in my publish folder.

When I double click the exe contained in that folder nothing happens. No errors or anything. It just does nothing.

All of the files are exactly the same as the ones in the publish folder in my dev project.

Can someone please help me figure out how to make a good windows installer for this app?

**Comment by amaitland** [comment]
> I tried a ClickOnce and it installs without errors including the .NET core 3.1

Which set of packages are you using? The older ones or the `.Net Core` specific ones (see https://github.com/cefsharp/CefSharp/issues/3574#issuecomment-869142293 for a list).

What version of `CefSharp` are you using?

> When I install that I end up with a C:\Program Files[companyname][appname] folder that has all of the same files contained in my publish folder.

Where are you installing the app? On your `dev` machine or a different machine? If different have you confirmed that `VC++ 2019` is installed?

**Comment by tpitman** [comment]
I am using the latest as of a couple days ago. I am using the .NET core 3.1 version.

I have tried installing on my dev computer and another computer. When I installed on the other computer it installed the .net core 3.1 as part of the install

It doesn't work on either my dev machine or the other one.

**Reply by amaitland** [reply]
> I am using the latest as of a couple days ago. I am using the .NET core 3.1 version.

Can you be specific please. `Latest` is relative. Which package exactly are you using? There are three.



> I have tried installing on my dev computer and another computer. When I installed on the other computer it installed the .net core 3.1 as part of the install

How exactly did you confirm the files are exactly the same?

**Reply by amaitland** [reply]
Did you check the event log to see if there were any errors reported there?

**Reply by tpitman** [reply]
You will have to give me a few minutes to start my VM back up to look up that information. I did not see any errors. I guess I am not sure they are exactly the same. It looked like they were the same files and I know the installer pulls them from that folder.

**Reply by tpitman** [reply]
CefSharp.Wpf.NETCore version 91.1.211.

I took a closer look at the files that got installed and found that in the win-x64\lib\netcoreapp3.1 folder in my publish folder of my project the files CefSharp.dll and ljwhost.dll are not in the Program Files install folder.

Any idea why those files would not get installed? My installer just says to put everything from the publish folder into the program files output folder.

Thank you for helping point this out, btw. Now I just need to know how to fix it.

**Reply by tpitman** [reply]
Also, I manually copied those files from the publish output folder into the same location in the Program Files folder and now it runs from there.

**Reply by tpitman** [reply]
Well, I need to get to bed. Thank you for helping. I look forward to any insight you can give as to why those 2 files didn't end up in my final installer. If you don't know how to do it in a MSI installer, but do know of how to make a ClickOnce work, I would be fine with that method as well.

**Reply by amaitland** [reply]
> I look forward to any insight you can give as to why those 2 files didn't end up in my final installer.

I don't know what the reason is. How exactly did you create the installer?

> but do know of how to make a ClickOnce work

Check that the files are included in the `ClickOnce` installer, there should be an option to view all included files, tick the show all files check box and make sure all the `bin/pak/dlls` files are included.
